File: 프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

vector<int> solution(vector<int> prices) {
    vector<int> answer;
    answer.resize(prices.size(), 0);
    stack<int> st;
    
    st.push(0);
    
    for(int i=1; i<prices.size(); i++){
        while(!st.empty() && prices[i] < prices[st.top()]){
            int j = st.top();
            st.pop();
            answer[j] = i- j;
        }
        st.push(i);
    }
    
    while(!st.empty()){
        int j = st.top();
        cout << j << "\n";
        st.pop();
        answer[j] = prices.size() - 1 - j;
    }
    
    
    return answer;
}

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        int[] answer = new int[prices.length];
        
        ArrayDeque<Integer> st = new ArrayDeque<>();
        ArrayList<Integer> list = new ArrayList<>();
        
        int cnt = 1;
        for(int i=0; i<prices.length; i++){
            while(!st.isEmpty() && prices[i] < prices[st.peekFirst()]){
                int top = st.peekFirst();
                st.pop();
                answer[top] = i-top;
            }
            st.push(i);
        }
        
        while(!st.isEmpty()){
                int top = st.peekFirst();
                st.pop();
                answer[top] = prices.length - top -1;
        }
        
        return answer;
    }
}